### PR TITLE
(DOCSP-3237) documented endpoints vs. programmatic discovery tool

### PR DESCRIPTION
**Pull request by danielborowski**
Thoughts and feedback.

### CODE REVIEW

File: tools/atlas-api-automation/differ.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        for route in data:
+            route_path = route['path']
+            # if this route is part of atlas
+            if route_path.find('atlas') != -1:
+                route_path = route_path.replace('/api/atlas/v1.0', '')
+                route_path = route_path.replace('/api/{app:public|atlas}/v1.0', '')
+                route_path = re.sub('{.+?}', '{*}', route_path)
+                if route_path != '':
+                    route_string = route['method'] + ' ' + route_path
+                    # pp.pprint(route_string)
```
Comment (i80and) [inline tools/atlas-api-automation/differ.py]: Generally avoid commented-out code like this; it tends to rot

File: tools/atlas-api-automation/differ.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                        atlas_endpoints_diff_from_docs.append(route_string)
+
+    # get whitelist of endpoints that we've deemed OK and do not include in output
+    with open(WHITELIST_FILE) as whitelist:
+        SKIP_ENDPOINTS = whitelist.read().splitlines()
+
+    # save the diff list to a file
+    with open(OUTPUT_FILE, 'w') as out_file:
+        written_to_output = 0
+        for endpoint in atlas_endpoints_diff_from_docs:
```
Comment (i80and) [inline tools/atlas-api-automation/differ.py]: Once you have the allow list created before creating `atlas_endpoints_diff_from_docs`, you can apply the allow list while creating it, and get rid of this loop:

`out_file.write('\n'.join(atlas_endpoints_diff_from_docs)`

And `written_to_output` becomes just `len(atlas_endpoints_diff_from_docs)`

File: tools/atlas-api-automation/differ.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        exit()
+
+    ATLAS_API_REFERENCE = args[0] + '/source/reference/api/'
+
+    # cloud-docs dir provided must contain api files
+    if not isdir(ATLAS_API_REFERENCE):
+        err = '''\nThe path: \n{apiref}\ndoes not contain a list of reference api files for atlas.
+        '''.format(apiref=ATLAS_API_REFERENCE)
+        print(err)
+        exit()
```
Comment (i80and) [inline tools/atlas-api-automation/differ.py]: `exit(1)` to indicate error

File: tools/atlas-api-automation/differ.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        with open(ATLAS_API_REFERENCE + api_file) as file_contents:
+            regex = re.search(ROUTE_REGEX, file_contents.read())
+            # if API route was found in text, replace base url and params with wildcard chars
+            if regex:
+                route_string = re.sub(r'(\/?api\/atlas\/v1\.0)', '', regex.group())
+                route_string = re.sub(r'{.+?}', '{*}', route_string)
+                route_string = route_string.replace('{*}:{*}', '{*}')
+                # some routes end with a trailing slash and so correct routes won't match
+                # e.g. 'GET /orgs/{*}/invoices' will not match 'GET /orgs/{*}/invoices/'
+                if route_string[-1] == '/':
```
Comment (i80and) [inline tools/atlas-api-automation/differ.py]: `if route_string.endswith('/'):`

File: tools/atlas-api-automation/differ.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    api_files = [f for f in listdir(ATLAS_API_REFERENCE) if isfile(join(ATLAS_API_REFERENCE, f))]
+
+    # will do a diff on these lists to see what new endpoints have not been documented yet
+    atlas_endpoints_in_docs = {}
+    atlas_endpoints_diff_from_docs = []
+
+    # store endpoints we currently have for atlas docs in dictionary
+    for api_file in api_files:
+        with open(ATLAS_API_REFERENCE + api_file) as file_contents:
+            regex = re.search(ROUTE_REGEX, file_contents.read())
```
Comment (i80and) [inline tools/atlas-api-automation/differ.py]: Better variable name would be `match`

File: tools/atlas-api-automation/differ.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            # if API route was found in text, replace base url and params with wildcard chars
+            if regex:
```
Comment (i80and) [inline tools/atlas-api-automation/differ.py]: I'd flip this condition to lower the indentation level:
```
if not match:
    continue
```

File: tools/atlas-api-automation/differ.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                # pp.pprint(route_string)
+                atlas_endpoints_in_docs[route_string] = True
+
+    # get json data of latest atlas code
+    with open(ATLAS_JSON_FILE) as atlas_json:
+        data = json.load(atlas_json)
+        for route in data:
+            route_path = route['path']
+            # if this route is part of atlas
+            if route_path.find('atlas') != -1:
```
Comment (i80and) [inline tools/atlas-api-automation/differ.py]: Flip this condition to lower the indentation level:
```
if 'atlas' not in route_path:
    continue
```

File: tools/atlas-api-automation/differ.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+# this is the regex that will look in the contents of atlas reference
+# api files and try to find the endpoints
+# e.g.  in "/source/reference/api/process-measurements.txt" it will match the line:
+#       GET api/atlas/v1.0/groups/{GROUP-ID}/processes/{HOST}:{PORT}/measurements
+ROUTE_REGEX = r'(?:POST|GET|DELETE|PATCH) (?:\/.*|api.*)'
+
+
+def main(args: List[str]) -> None:
+
```
Comment (i80and) [inline tools/atlas-api-automation/differ.py]: Nix this empty line

File: tools/atlas-api-automation/differ.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            route_string = re.sub(r'{.+?}', '{*}', route_string)
+            route_string = route_string.replace('{*}:{*}', '{*}')
+            # some routes end with a trailing slash and so correct routes won't match
+            # e.g. 'GET /orgs/{*}/invoices' will not match 'GET /orgs/{*}/invoices/'
+            if route_string.endswith('/'):
+                route_string = route_string[:-1]
+            atlas_endpoints_in_docs[route_string] = True
+
+    # get whitelist of endpoints that we've deemed OK and do not include in output
+    skip_endpoints = None
```
Comment (i80and) [inline tools/atlas-api-automation/differ.py]: I THINK you don't need this line, because it's defined below? I might be lying

**Comment by danielborowski** [comment]
@i80and fixed

**Pull request merged**